src/context.rs

```diff
@@ -0,0 +1,213 @@
+use futures::future::{pending, BoxFuture, FutureExt, Shared};
+use tokio::time::Duration;
+
+pub struct CancelHandle {
+    cancel_fn: Box<dyn FnOnce()>,
+}
+
+impl CancelHandle {
+    pub fn cancel(self) {
+        (self.cancel_fn)();
+    }
+}
+
+#[derive(Clone, Debug)]
+pub struct Context {
+    pub done: Shared<BoxFuture<'static, ()>>,
+}
+
+impl Context {
+    pub fn new() -> Self {
+        Self {
+            done: pending().boxed().shared(),
+        }
+    }
+
+    pub fn with_cancel(&self) -> (Self, CancelHandle) {
+        let (done0, canceler) = futures::future::abortable(self.done.clone());
+        let done = done0.map(|_| ()).boxed().shared();
+        let cancel_fn = Box::new(move || { canceler.abort() });
+        (Self { done }, CancelHandle { cancel_fn })
+    }
+
+    pub fn with_timeout(&self, timeout: Duration) -> (Self, CancelHandle) {
+        let done0 = tokio::time::timeout(timeout, self.done.clone());
+        let (done1, canceler) = futures::future::abortable(done0);
```

Comment:
FYI, in Rust with `let` you can just rebind the same name multiple times and it's idiomatic to do so. It just shadows the old binding, which is fine to do since you are moving the `done*` values each time you use them.

Reply:
Gotcha, this is the Haskeller in me doing the suffix indexes